### PR TITLE
Merge firestore bundles implementation without exposing public APIs

### PR DESCRIPTION
**Pull request by wu-hui**

### CODE REVIEW

File: .changeset/lemon-steaks-draw.md
Comment (schmidt-sebastian) [inline .changeset/lemon-steaks-draw.md:1]: Please add a changeset entry to describe the size increase.

File: packages/firestore-types/index.d.ts
Comment (schmidt-sebastian) [inline packages/firestore-types/index.d.ts:125]: For now, we should remove this from this file.

File: packages/firestore/exp-types/index.d.ts
Comment (schmidt-sebastian) [inline packages/firestore/exp-types/index.d.ts:545]: These can stay. You can also just make this feature public in firestore-exp.

File: packages/firestore/exp/index.ts
Comment (schmidt-sebastian) [inline packages/firestore/exp/index.ts:34]: Feel free to make these APIs public in firestore-exp.

File: packages/firestore/exp/src/api/database.ts
Comment (schmidt-sebastian) [inline packages/firestore/exp/src/api/database.ts:507]: FWIW, I have lately followed Gil's suggestions more strictly and prefixed the free standing APIs with the class name.

That would mean s/loadBundleSyncEngine/syncEngineLoadBundle/

File: packages/firestore/exp/src/api/database.ts
Comment (schmidt-sebastian) [inline packages/firestore/exp/src/api/database.ts:510]: This pattern means that the resulting user task never fails.  If you want to log an error, you should catch and rethrow here. A better way might be to log an error when the error occurs.
Note that you also should not return this task to the AsyncQueue, as it will cause it to fail. Instead, just return an empty Promise (which is the default behavior for async functions, so you can just remove `return`).

File: packages/firestore/exp/src/api/database.ts
Comment (schmidt-sebastian) [inline packages/firestore/exp/src/api/database.ts:526]: Let's fix this before merging.

File: packages/firestore/exp/test/shim.ts
Comment (schmidt-sebastian) [inline packages/firestore/exp/test/shim.ts:175]: Let's fix this before merging.

File: packages/firestore/rollup.shared.js
Comment (schmidt-sebastian) [inline packages/firestore/rollup.shared.js:133]: Should we just add all JS keywords here? This will fail with `if` and `of` next and it is really tough to figure out these failures.

File: packages/firestore/src/core/firestore_client.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/core/firestore_client.ts:599]: Same comment as above.

File: packages/firestore/src/core/firestore_client.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/core/firestore_client.ts:605]: This should happen on the AsyncQueue.

File: packages/firestore/src/local/local_store.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_store.ts:1281]: This doesn't need to be exported.

File: packages/firestore/src/local/shared_client_state_schema.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/local/shared_client_state_schema.ts:122]:
```suggestion
//     firestore_bundle_loaded_<persistenceKey>
export const BUNDLE_LOADED_KEY_PREFIX = 'firestore_bundle_loaded';
```

File: packages/firestore/test/integration/api_internal/bundle.test.ts
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api_internal/bundle.test.ts:34]: I would consider this blocking as this change will ensure that this feature works in the minified Browser build and is also required to ensure that the firestore-exp code you added is tested.

File: packages/firestore/src/core/firestore_client.ts
Comment (wu-hui) [reply packages/firestore/src/core/firestore_client.ts]: This looks different now, but should be fixed.

File: packages/firestore/src/core/firestore_client.ts
Comment (wu-hui) [reply packages/firestore/src/core/firestore_client.ts]: Done.

File: packages/firestore/src/local/shared_client_state_schema.ts
Comment (wu-hui) [reply packages/firestore/src/local/shared_client_state_schema.ts:122]: Done.

File: packages/firestore/test/integration/api_internal/bundle.test.ts
Comment (wu-hui) [reply packages/firestore/test/integration/api_internal/bundle.test.ts]: Done.

File: packages/firestore/exp/src/api/database.ts
Comment (wu-hui) [reply packages/firestore/exp/src/api/database.ts]: Done.

File: packages/firestore/exp/src/api/database.ts
Comment (wu-hui) [reply packages/firestore/exp/src/api/database.ts]: I think this is fixed, although this particular piece of code no longer exists.

File: packages/firestore/exp/src/api/database.ts
Comment (wu-hui) [reply packages/firestore/exp/src/api/database.ts]: Done.

File: packages/firestore/exp/test/shim.ts
Comment (wu-hui) [reply packages/firestore/exp/test/shim.ts]: Done.

File: packages/firestore/rollup.shared.js
Comment (wu-hui) [reply packages/firestore/rollup.shared.js:133]: Done.

**Review by wu-hui — COMMENTED** [review]
This has changed quite a bit on the api/firestore_client level since last time you reviewed it. 

Also, this is no longer release but hide it from public API, it is release bundles as free functions. I will hold it until we got all approvals and ready to launch.

File: packages/firestore/src/api/database.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts]: You can keep these functions as long as they are not referenced.

File: packages/firestore/src/core/firestore_client.ts
Comment (schmidt-sebastian) [reply packages/firestore/src/core/firestore_client.ts:665]: Same just below.

File: packages/firestore/src/core/sync_engine.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts:1556]: Could this potentially be inlined? Or did we not want to make `syncEngineLoadBundle` async?

**Review by schmidt-sebastian — COMMENTED** [review]
Looks there there are still some stray API bits, but otherwise good. Since the size tooling seems to be broken at the moment, can you provide a rough number on the size increase?

**Review by wu-hui — COMMENTED** [review]
firebase-firestore.js 283361 -> 286235
firebase-firestore.memory.js 221571 -> 224766

File: packages/firestore/src/api/database.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts:434]: We need to figure out where to call "ensureClientConfigured". It might depend on how we expose this function in the end.

**Review by schmidt-sebastian — APPROVED** [review]
Wheee!

**Pull request merged**
